src/pages/Swap/LimitOrderBox/components/LimitOrderForm/LimitOrderForm.tsx

```diff
@@ -0,0 +1,525 @@
+        nextLimitPriceFloat !== Infinity ? nextLimitPriceFloat.toFixed(6) : '0',
+        currency.decimals
+      ).toString()
+    } else {
+      nextLimitPriceFloat = sellTokenAmountFloat / nextBuyAmountFloat
+      // divide the buy amount by the new price
+      newLimitOrder.limitPrice = JSBI.multiply(newBuyTokenAmount.raw, buyTokenAmount.raw).toString()
+    }
+
+    // Update state
+    setFormattedBuyAmount(nextBuyAmountFormatted) // update the token amount input
+    setFormattedLimitPrice(nextLimitPriceFloat.toString()) // update the token amount input
+    setLimitOrder(newLimitOrder)
+  }
+
+  const handleInputOnChange = (token: Token, handleAmountChange: Function) => (formattedValue: string) => {
+    const amountFormatted = formattedValue.trim() === '' ? '0' : formattedValue
+    const amountWei = parseUnits(formattedValue, token.decimals).toString()
+    handleAmountChange({
+      currency: token as Token,
+      amountWei,
+      amountFormatted,
+    })
+  }
+
+  const handleCurrencySelect =
+    (prevTokenAmount: TokenAmount, handleCurrencyAmountChange: Function) => (currency: Currency) => {
+      const amountWei = prevTokenAmount?.raw
+        ? prevTokenAmount.raw.toString()
+        : formattedBuyAmount
+        ? parseUnits(formattedBuyAmount, prevTokenAmount?.currency?.decimals).toString()
+        : '0' // use 0 if no buy currency amount is set
+      handleCurrencyAmountChange({ currency, amountWei, amountFormatted: formattedBuyAmount })
+    }
+
+  const [marketPrices, setMarketPrices] = useState<MarketPrices>({ buy: 0, sell: 0 })
+
+  const getMarketPrices = useCallback(async () => {
+    const signer = provider.getSigner()
+    if (limitOrder.buyToken && limitOrder.sellToken) {
+      const order = JSON.parse(JSON.stringify(limitOrder))
+
+      const token = limitOrder.kind === LimitOrderKind.SELL ? sellTokenAmount : buyTokenAmount
+
+      order.sellAmount = parseUnits('10', token.currency.decimals).toString()
+
+      const cowQuote = await getQuote({
+        chainId,
+        signer,
+        order: { ...order, expiresAt: dayjs().add(expiresIn, expiresInUnit).unix() },
+      })
+
+      if (cowQuote) {
+        const {
+          quote: { buyAmount, sellAmount },
+        } = cowQuote
+
+        if (limitOrder.kind === LimitOrderKind.SELL)
+          setMarketPrices(marketPrice => ({
+            ...marketPrice,
+            buy: formatMarketPrice(buyAmount, buyTokenAmount.currency.decimals),
+          }))
+        else
+          setMarketPrices(marketPrice => ({
+            ...marketPrice,
+            sell: formatMarketPrice(sellAmount, sellTokenAmount.currency.decimals),
+          }))
+      }
+    }
+  }, [buyTokenAmount, chainId, expiresIn, expiresInUnit, limitOrder, provider, sellTokenAmount])
+
+  useEffect(() => {
+    getMarketPrices()
+  }, [limitOrder.kind, getMarketPrices])
+
+  useEffect(() => {
+    setMarketPrices({ buy: 0, sell: 0 })
+  }, [limitOrder.sellToken, limitOrder.buyToken])
+
```

Comment:
I agree with Bertoetti on this point there is wayyyy to much logic here and its very unreadable almost 400 lines of logic. This should be broken down further..

Reply:
We can do this in a new PR if it makes sense. But we must do it now or it will be forgotten.